### DISCUSSION TITLE
Setting up local development help?

### DISCUSSION THREAD

**Discussion by arovillard** [vicwomg/pikaraoke#272] [Q&A]
I'm a developer but I have only ever worked with react and JS. I was hoping to do some changes and use this project to learn more about developing this kind of apps.

I was able to ssh into pi from my mac and load the pikaraoke in VSstudio, but the changes I'm making in this directory are not being reflected in the running program, even if I restart it. I was also able to spot the --development-mode flag but the new server port it opens just opens nothing for me.

Could someone share some pointers on how I can do local development and have my changes be reflected in the app when I run it? I believe the dev mode flag should allow me to do this but it is unclear how it is supposed to work.

Cheers

**Comment by vicwomg** [comment]
First of all, which files are you trying to edit? `--development-mode` only reflects changes you make the html files under `/templates/` I don't believe any changes to the python files will be reflected unless you restart.

As for why the changes don't take effect when you restart, that is more puzzling to me. My first inclination is that perhaps you're not changing what you think you are. More specifics would be useful to know.

**Comment by arovillard** [comment]
Thanks for getting back to me, so what I wanted to do was to play around and maybe build some other functionality. So the files I was changing were the ones from the pikaraoke directory (the ones that get updated when you do a git pull). I'm using the pre-built image you provided with the only difference that I had to do a git pull and ran the "./setup-pi.sh" script again.

I was stopping the app and making visual changes to the app (the one you can access through a browser) but not seeing changes. I tried stopping the app and starting it again (sudo python3 app.py) but the changes still did not update. I was not using the --dev flag FYI.

What should be the correct development process for making changes to the app and seeing them reflected? the changes I was making where in the templates (html files). Is the app being served from a build folder I am not seeing?

Any help would be greatly appreciated, there are a lot of devs out there that would probably love to help out as well. I was even planning on building a react app (as I would feel a lot more familiarized) and use the API endpoints to control and fetch the data from the app.

Cheers!

**Comment by vicwomg** [comment]
If I had to redo it, I would do it in react. Though last I checked performance with builds on raspberry pis can be disappointing.

There is no build folder for flask. Restarting should definitely show the changes, and I have never heard otherwise from other contributors. So again, not sure what's going wrong there.  Is it possible you're inadvertently editing a copy of the code that's running?

**Comment by vicwomg** [comment]
I actually just had a thought. If you're using the pre-built image, it auto launches on start from /etc/rc.local script. Perhaps its still running in memory and that's what you're looking at in your browser. 

Make sure to kill that first `ps aux  |  grep python` will show you the processes. Then when you're developing you can manually handle starting/restarting the app and using dev mode as you like.

**Reply by arovillard** [reply]
Oh this could definitely be the case. And it would make sense too.
When I added the dev flag it created a new port 127.### but nothing was being served there. After I kill the process, would I be starting the app the same way (sudo python3 app.py)? or is there another way I should be starting the project?

**Reply by vicwomg** [reply]
Yes, that's the way.

**Reply by arovillard** [reply]
ok I can confirm that I can see changes when I restart. The only thing I need to figure out is why using "sudo python3 app.py --developer-mode" flag is not working. 
I get this on the console:
```
WARNING:root:Splash screen is disabled in developer mode due to main thread conflicts
 * Serving Flask app "app" (lazy loading)
 * Environment: production
   WARNING: This is a development server. Do not use it in a production deployment.
   Use a production WSGI server instead.
 * Debug mode: on
INFO:werkzeug: * Running on http://127.0.0.1:5555/ (Press CTRL+C to quit)
INFO:werkzeug: * Restarting with stat
pygame 1.9.6
Hello from the pygame community. https://www.pygame.org/contribute.html
WARNING:root:Splash screen is disabled in developer mode due to main thread conflicts
WARNING:werkzeug: * Debugger is active!
INFO:werkzeug: * Debugger PIN: 363-722-729
```

Would you know what is going on? going to http://127.0.0.1:5555/ shows nothing.

I appreciate all your help!

**Reply by vicwomg** [reply]
`127.0.0.1` is not a real IP, that's basically a shorthand for `localhost`. So it will only work if you open up a browser from the pi itself.

Use the same QR code from non-dev mode. It will work for both.
